torchtune/modules/vision_transformer.py

```diff
@@ -0,0 +1,334 @@
+    was tile-cropped, then you need to use tile positional embeddings:
+
+    - token_pos_embedding (tiled): ``torchtune.models.clip._position_embeddings.TiledTokenPositionalEmbedding``
+    - pre_tile_pos_embed: ``torchtune.models.clip._position_embeddings.TilePositionalEmbedding``
+    - post_tile_pos_embed: ``torchtune.models.clip._position_embeddings.TilePositionalEmbedding``
+
+    Otherwise, pre and post tile_pos_embed should be None. For the tokens, all you need is a simple
+    token positional embedding:
+
+    - token_pos_embedding: ``torchtune.models.clip._position_embeddings.TokenPositionalEmbedding``
+
+    All images will be considered as a stack of tiles, even if your image was not tile-cropped. In such cases,
+    your image would be composed of a single tile.
+
+    Args:
+        patch_grid_size (int): The side of a squared grid that represents how many
+             patches are in one tile, i.e. tile_size // patch_size.
+        num_layers (int): The number of transformer layers.
+        layer (nn.Module): The transformer layer module.
+        token_pos_embedding (nn.Module): The token positional embedding module.
+        pre_tile_pos_embed (Optional[nn.Module]): The pre-tile positional embedding module. It should be
+            None if your image was not tile-cropped in advance.
+        post_tile_pos_embed (Optional[nn.Module]): The post-tile positional embedding module. It should be
+            None if your image was not tile-cropped in advance.
+        cls_projection (Optional[nn.Module]): The CLS projection module. It should take an input tensor
+            of shape (bsz * n_tiles, n_tokens, embed_dim) and output a tensor of shape
+            (bsz * n_tiles, cls_output_dim). If provided, only the CLS token projection will be
+            outputted, instead of all tokens.
+        indices_return_hidden (Optional[List[int]]): The indices of hidden layers to return.
+            If provided, it will return the intermediate results of the transformer layers
+            before they go through a next layer. For example, indices_return_hidden = [0, 3] will
+            return the tokens before they go through the first and fourth layers.
+        patch_size (int): The size of each patch. Used to divide the tiles into patches.
+            E.g. for patch_size = 40, a tile of shape (400, 400) will have 10x10 grid of patches
+            with shape (40, 40) each.
+        embed_dim (int): The dimensionality of each patch embedding (token).
+        in_channels (int): The number of image input channels.
+    """
+
+    def __init__(
+        self,
+        patch_grid_size: int,
+        num_layers: int,
+        layer: nn.Module,
+        token_pos_embedding: nn.Module,
+        pre_tile_pos_embed: Optional[nn.Module] = None,
+        post_tile_pos_embed: Optional[nn.Module] = None,
+        cls_projection: Optional[nn.Module] = None,
+        indices_return_hidden: Optional[List[int]] = None,
```

Comment:
Two questions about this argument: (1) do we need that much configurability on which hidden states are returned? Why not just a bool? (2) Why is it defined in init and not in forward? Personally I think forward is the more natural place for it

Reply:
1) a larger model has more layers, so a 3b model will return some indices, and 1b will return others, for example. I think it also allows users to play with it: should it return every N, every M or all of the layers?

2) good question! Both would work. I guess that in the forward the user has more flexibility, but then they have to keep track of one more parameter during inference.

Do you think that this is something that they would change with every call, or its the type of parameter that has to be defined only once, and it is ok to be in __init__?

personally i prefer it in the init, so save complexity in the inference, but either way works for me. Do you think it is worth the change?

Reply:
So I think it depends on how people will actually use this. 

First, a question: My impression was that the "cross-attention every Nth layer" of Flamingo is just using the final visual embeddings and not any intermediate states. Is that the right understanding? 

Assuming so:

(1) makes sense, though I bias towards only adding additional configurability if we have a clear expectation that people will use it. So unless there's a clear case here I would leave it as a bool for an initial version. (But if the intermediate states are used in every Nth cross-attention layer, I think it's fine to leave this as is.)

For (2) my mental model here is `return_hidden_states`, which is typically a forward argument in all the places I've seen. So I would recommend to move it to forward just so that we are consistent with users' expectations (separately as some of our `TransformerDecoder` refactor work we will be doing something similar). It's kinda tied in with (1) though, like if we do wind up going with a list of integers then this is a less compelling argument.

But I can imagine a case where the same encoder gets called twice on a single forward pass of the model.. say we have a batch containing images and videos and we want to do some extra aggregation of videos using their hidden states, but not images. In that case, moving the param into forward is actually practically useful. (Admittedly a somewhat contrived example, but I have seen other cases in multimodal land where the same encoder is used multiple times in a single forward pass, once with pooling over intermediate states and once without.)

Reply:
Evan and I chatted, and given how it is used by the adapter, we will keep it as it is. Thanks for the comments!!